### ISSUE TITLE
Exception when moving fast between images that cannot be displayed

### ISSUE THREAD

**Issue by Urmel** [nemethviktor/GeoTagNinja#16]
Hi!

first of all thanks for taking on the task of starting this project. I've been using GeoSetter for quite some time, but as it now is almost completely useless due to the latest Google Api update, was looking for an alternative. And your project for sure looks like a good one!

### Setting

So I started with the latest greatest you provided in this issue: https://github.com/nemethviktor/GeoTagNinja/issues/15.

Running on a German Win10 machine.

### Error

I have a folder with files that the Ninja can display (the upper ones) and 4 images the Ninja cannot display (the lower ones; issue https://github.com/nemethviktor/GeoTagNinja/issues/17):
![grafik](https://user-images.githubusercontent.com/1786902/203157692-683730b9-7edc-44da-b5e7-512e3763e762.png)

When I quickly curser up and down through the images that Ninja can display, everything is fine. The application takes some time to update the preview and it works it way throught the long list of keypresses after I finished - but nothing crashes.

When I do this with the images in the lower part - i.e. quickly cursor up and down - I get this exception:

> System.ArgumentException: Ungültiger Parameter.
>    bei System.Drawing.Bitmap..ctor(String filename)
>    bei GeoTagNinja.HelperStatic.<LvwItemCreatePreview>d__33.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.FrmMainApp.<lvw_FileList_KeyUp>d__58.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)

I have no clue, why it's in German (ok, my windows is, but....). So...
- Ungültiger Parameter -> "illegal argument" / parameter
- "Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde" seems like an automatically translated thing. Sounds like "end of stack trace for observed scope"

Note that this *will not* happen, when I cursor slowly through them ...

**Comment by nemethviktor** [comment]
Hey, thanks for the report.

Yeah, amusingly enough I've bumped into that one myself recently! :D
Am working on a bigger update that will hopefully solve this plus make a few things a little faster. If all goes to plan I can upload a development version (like you've found the one in #15 ) today or tomorrow and then you can have a look. 
Will keep you posted.

**Comment by Urmel** [comment]
Hi @nemethviktor,

I used the binary from ~an hour ago, issue https://github.com/nemethviktor/GeoTagNinja/issues/19 (hope that was correct...). With that I did some fast cursoring up and down again:
- With images that can be displayed in the preview pane, it works.
- With images that cannot be displayed, **the issue still persists** (but it seems that it takes a bit longer for it to show up)

Some more observations:
- In general when switching the image, first, the image is shown and then the map blinks to show the new position
- For the non-displayable images, the time between cursering to an image and the resulting map blink takes longer than for a displayable image
- For the displayable images, when I cursor through them, the cursor moves much slower through the table than for the non-displayable images

Exception shown - test no 1:
> System.ArgumentException: Ungültiger Parameter.
>    bei System.Drawing.Bitmap..ctor(String filename)
>    bei GeoTagNinja.HelperStatic.<LvwItemCreatePreview>d__41.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.HelperStatic.<LvwItemClickNavigate>d__39.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.FrmMainApp.<lvw_FileList_KeyUp>d__61.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)

Exception shown test no 2:

> System.ArgumentException: Ungültiger Parameter.
>    bei System.Drawing.Bitmap..ctor(String filename)
>    bei GeoTagNinja.HelperStatic.<LvwItemCreatePreview>d__41.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.HelperStatic.<LvwItemClickNavigate>d__39.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.FrmMainApp.<lvw_FileList_KeyUp>d__61.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)

Exception shown test no 3:

> System.ArgumentException: Ungültiger Parameter.
>    bei System.Drawing.Bitmap..ctor(String filename)
>    bei GeoTagNinja.HelperStatic.<LvwItemCreatePreview>d__41.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.TaskAwaiter.ThrowForNonSuccess(Task task)
>    bei System.Runtime.CompilerServices.TaskAwaiter.HandleNonSuccessAndDebuggerNotification(Task task)
>    bei GeoTagNinja.FrmMainApp.<lvw_FileList_KeyUp>d__61.MoveNext()
> --- Ende der Stapelüberwachung vom vorhergehenden Ort, an dem die Ausnahme ausgelöst wurde ---
>    bei System.Runtime.CompilerServices.AsyncMethodBuilderCore.<>c.<ThrowAsync>b__6_0(Object state)

**Comment by nemethviktor** [comment]
Re: speed/or the lack of: The logic for preview creation is (as of now, which is to say this logic includes the change from the other thread)
- if extension = jpg then use windows' own thing, This is fast but doesn't really support too many formats.
- else try to generate via exiftool. this is slow
-- if ET-generated preview exists use that
-- else try windows again regardless of extension
--- if that works, yayy
--- else there is no preview.

if you're still around i just uploaded another one (timestamp of the file is 2256pm uk time filesize 16,458,240 bytes), see if you still get the problem. i tried wrapping one of the methods in a try/catch so it doesn't crash...supposedly. build number same as before. (just re-download, easiest.)

**Comment by nemethviktor** [comment]
Aaaaactualllly. Thinking out loud helps. That logic is stupid.
Same link as above, TS 2306 UK time.
Logic updated:
- try Windows' own preview creation regardless of the image extension. 
- if this fails try exiftool.
- if that fails there is no output.

**Comment by pbranly** [comment]
Hi
I would like as well to test the latest versions but I am not sure it is as the download link does not give any choice 
Is it always the latest one ? 
phil

**Comment by pbranly** [comment]
Ok thanks
Yes I saw the modification …. For sure ..
I have to work on the French translations as soon as I have some time. And there is an issue with the available length of the sentence .
If you don’t mind I will try to send you the modified  French SQLite file
Good job
Phil

**Comment by nemethviktor** [comment]
Sure that's ok

<http://uk.linkedin.com/in/viktornemeth>
<https://www.instagram.com/alien_bunny_ninja/>

**Comment by nemethviktor** [comment]
Hi both, 
Just to make testing a little less confusing, going forward the About Box will have not only the version number but also a build datetime. This should help avoiding the problem where I publish more than 1 build a day to test and then it's difficult to guess if we're all on the same page or not.

**Comment by nemethviktor** [comment]
@Urmel  can you have a look pls at the latest version (same link as above)
![image](https://user-images.githubusercontent.com/58767235/203657055-50248645-f5a3-4000-9248-49c086e3f79b.png)

I think that should settle both your open bug tickets. Once everyone's happy I'll do a release and then work on the feature requests.

**Comment by Urmel** [comment]
Hi @nemethviktor,

I tested using (can we make that text copy-able?):
![grafik](https://user-images.githubusercontent.com/1786902/203866694-2d717317-86fa-4a09-bd5f-738e2e682e21.png)

I did the usual cursoring up and down:
- Cursoring up and down in the "regular" files (the ones that were already working - likely had a working preview graphic) was as slow as before
- Cursoring through the files that were not working before went "super fast" - I'm not able to build up a relevant delay between stoping cursoring versus the last screen refresh. In contrast to "previously working images", where there is multi second delay after doing some wild cursoring.

No more exceptions!
**So for me, this fix solves the problem!**

Thank you!!
